# Modelling: Freight Rate Prediction

This notebook documents the full modelling process end-to-end:
1. **Model Comparison** -- ExtraTrees, RandomForest, HGB variants benchmarked on a held-out fold
2. **Why DollarHGBRegressor wins** -- ablations isolating distance-weighting, Duan smearing, native categoricals
3. **Forward-chaining CV** -- 4-fold time-series evaluation + unseen-city stress test
4. **Market index ablation** -- feature importance / sensitivity analysis
5. **Final fit & deliverables** -- production pipeline written to `output/pipeline.joblib`

Core logic lives in `train.py` / `preprocessing.py` / `modeling.py` -- this notebook **calls** those modules so `python train.py` and the notebook cannot drift.

In [1]:
import sys, json, warnings
from pathlib import Path
warnings.filterwarnings('ignore')
sys.path.insert(0, str(Path('..').resolve()))
import numpy as np, pandas as pd, matplotlib.pyplot as plt, seaborn as sns
import train as T
from preprocessing import MODEL_FEATURES, CAT_FEATURES, NUM_FEATURES
from modeling import DollarHGBRegressor
from sklearn.ensemble import ExtraTreesRegressor, RandomForestRegressor, HistGradientBoostingRegressor
from score import slice_report
DATA, RESULTS = Path('../data'), Path('results/modelling')
RESULTS.mkdir(parents=True, exist_ok=True)
train = pd.read_csv(DATA/'train_test.csv', parse_dates=['date'])
validation = pd.read_csv(DATA/'validation.csv', parse_dates=['date'])
print('train', train.shape, train.date.min().date(), '->', train.date.max().date(), '| validation', validation.shape)
print('MODEL_FEATURES:', len(MODEL_FEATURES), '| CAT:', CAT_FEATURES)

train (48113, 14) 2025-01-01 -> 2025-10-31 | validation (12000, 13)
MODEL_FEATURES: 21 | CAT: ['pickup_cat', 'delivery_cat', 'equipment_cat']


## 1. Cleaning (training-only corrupt label flags)
Sign-flipped weights fixed via `basic_clean`; `6*MAD` robust quadratic rule flags ~1.4% corrupt labels from *training folds only*. Inference path never drops rows.

In [2]:
train_raw = T.basic_clean(train)
train_raw['is_corrupt'] = T.flag_corrupt_labels(train_raw)
print(f"Corrupt labels flagged: {train_raw.is_corrupt.sum()} ({train_raw.is_corrupt.mean():.2%})")

Corrupt labels flagged: 677 (1.41%)


## 2. Inductive Imputation -- No Transduction
FORBIDDEN (leak): `X.groupby('date').median()` on validation/inference.  
PRODUCTION: `MarketImputer` maps frozen train lookups date->month->dow->global.  
Single-row inference == batch inference by construction.

In [3]:
# LEAK (do not use): X.loc[m,'market_index'] = X.loc[m,'date'].map(X.groupby('date').market_index.median())
tr_ = train_raw[train_raw.date <= '2025-07-31']
va_ = train_raw[(train_raw.date >= '2025-08-01') & (train_raw.date <= '2025-08-31')].head(200).copy()
va_.loc[va_.sample(20, random_state=0).index, 'market_index'] = float('nan')
prep = T.fit_preprocessors(tr_)
batch = T.transform(va_, prep)
single = T.transform(va_.head(1), prep)
assert np.isclose(batch.market_index.iloc[0], single.market_index.iloc[0]), 'parity broken'
print('single-row vs batch parity: OK; first-row market =', round(float(single.market_index.iloc[0]),4))
print('fitted ladder sizes: date=%d month=%d dow=%d global=%.3f' % (
    len(prep['_m'].date_map_), len(prep['_m'].month_map_), len(prep['_m'].dow_map_), prep['_m'].global_))

single-row vs batch parity: OK; first-row market = 1.5451
fitted ladder sizes: date=212 month=10 dow=7 global=1.517


## 3. Model Comparison (Fold 2: train <= Jul 31, validate Aug-Sep)

All models are trained on the **same cleaned, preprocessed features** with the same 6*MAD outlier exclusion
and `log(rate/mile)` target. Results sorted by `clean_mae` ascending -- lower is better.

**Design choices being compared:**
- `sample_weight=distance` (dollar-space calibration vs unweighted)
- Duan smearing correction `S = mean(exp(resid))` (bias correction after log-transform)
- Native HGB categoricals for `pickup_cat / delivery_cat / equipment_cat` via `from_dtype`

> Note: ExtraTrees and RandomForest use `n_estimators=100` for runtime efficiency;
> DollarHGB uses `max_iter=300` (its standard production setting).

In [4]:
# Fold 2 setup
trf = train_raw[train_raw.date <= '2025-07-31']
vaf = train_raw[(train_raw.date >= '2025-08-01') & (train_raw.date <= '2025-09-30')]
print(f'TR rows: {len(trf)} | VA rows: {len(vaf)}')

prep = T.fit_preprocessors(trf)
tr_t = T.transform(trf, prep)
va_t = T.transform(vaf, prep)

# Outlier mask fitted on training fold ONLY (no holdout peeking)
coef, med, mad = T._flag_params(trf)
tr_mask = T._apply_flag(tr_t, coef, med, mad)
va_mask = T._apply_flag(va_t, coef, med, mad)
tr_clean = tr_t[~tr_mask]

w = tr_clean['distance'].to_numpy()                          # sample_weight=distance
y = np.log(tr_clean['posted_rate'] / tr_clean['distance'])   # target: log(rpm)

print('\nRunning 6-model benchmark on Fold 2 (Aug-Sep holdout)...')
rows = []

# 1. ExtraTrees sqrt features -- original baseline (no distance weighting)
print('  [1/6] ExtraTrees (sqrt, n=100) ..................... ', end='', flush=True)
et1 = ExtraTreesRegressor(n_estimators=100, min_samples_leaf=2, max_features='sqrt', n_jobs=-1, random_state=42)
et1.fit(tr_clean[NUM_FEATURES], y)
p1 = np.exp(et1.predict(va_t[NUM_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'ExtraTrees (sqrt, n=100)', **T.score(vaf['posted_rate'], p1, va_mask)})
print('done')

# 2. ExtraTrees + distance weighting
print('  [2/6] ExtraTrees (sqrt, distance-weighted, n=100) .. ', end='', flush=True)
et2 = ExtraTreesRegressor(n_estimators=100, min_samples_leaf=2, max_features='sqrt', n_jobs=-1, random_state=42)
et2.fit(tr_clean[NUM_FEATURES], y, sample_weight=w)
p2 = np.exp(et2.predict(va_t[NUM_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'ExtraTrees (sqrt, distance-weighted, n=100)', **T.score(vaf['posted_rate'], p2, va_mask)})
print('done')

# 3. RandomForest + distance weighting
print('  [3/6] RandomForest (sqrt, distance-weighted, n=100)  ', end='', flush=True)
rf = RandomForestRegressor(n_estimators=100, min_samples_leaf=2, max_features='sqrt', n_jobs=-1, random_state=42)
rf.fit(tr_clean[NUM_FEATURES], y, sample_weight=w)
p3 = np.exp(rf.predict(va_t[NUM_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'RandomForest (sqrt, distance-weighted, n=100)', **T.score(vaf['posted_rate'], p3, va_mask)})
print('done')

# 4. HGB plain -- no weighting, no native categoricals
print('  [4/6] HGB plain (no weighting, no cats) ........... ', end='', flush=True)
hgb_plain = HistGradientBoostingRegressor(max_iter=300, learning_rate=0.05, max_leaf_nodes=63, min_samples_leaf=20, random_state=42)
hgb_plain.fit(tr_clean[NUM_FEATURES], y)
p4 = np.exp(hgb_plain.predict(va_t[NUM_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'HGB plain (no weighting, no cats)', **T.score(vaf['posted_rate'], p4, va_mask)})
print('done')

# 5. HGB + distance weighting (no smearing, no native cats)
print('  [5/6] HGB + distance-weighted ...................... ', end='', flush=True)
hgb_w = HistGradientBoostingRegressor(max_iter=300, learning_rate=0.05, max_leaf_nodes=63, min_samples_leaf=20, random_state=42)
hgb_w.fit(tr_clean[NUM_FEATURES], y, sample_weight=w)
p5 = np.exp(hgb_w.predict(va_t[NUM_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'HGB + distance-weighted', **T.score(vaf['posted_rate'], p5, va_mask)})
print('done')

# 6. DollarHGBRegressor -- full production: weighting + Duan smearing + native cats
print('  [6/6] DollarHGBRegressor [PRODUCTION] .............. ', end='', flush=True)
dollar = DollarHGBRegressor(
    max_iter=300, max_leaf_nodes=63, learning_rate=0.05,
    min_samples_leaf=20, l2_regularization=1.0,
    categorical_features='from_dtype', random_state=42)
dollar.fit(tr_clean[MODEL_FEATURES], y, sample_weight=w)
p6 = np.exp(dollar.predict(va_t[MODEL_FEATURES])) * va_t['distance'].values
rows.append({'Model': 'DollarHGBRegressor (weighted + smearing + native cats) [PRODUCTION]',
             **T.score(vaf['posted_rate'], p6, va_mask)})
print('done\n')

bench = pd.DataFrame(rows).sort_values('clean_mae').reset_index(drop=True)
bench

TR rows: 33718 | VA rows: 9429

Running 6-model benchmark on Fold 2 (Aug-Sep holdout)...
  [1/6] ExtraTrees (sqrt, n=100) ..................... done
  [2/6] ExtraTrees (sqrt, distance-weighted, n=100) .. done
  [3/6] RandomForest (sqrt, distance-weighted, n=100)  done
  [4/6] HGB plain (no weighting, no cats) ........... done
  [5/6] HGB + distance-weighted ...................... done
  [6/6] DollarHGBRegressor [PRODUCTION] .............. done



,Model,clean_mae,clean_rmse,clean_mape_pct,all_row_mae
0,[BEST] DollarHGBRegressor (weighted + smearing + native cats) [PRODUCTION],39.28,57.93,1.69,90.08
1,HGB + distance-weighted,39.96,57.79,1.75,90.73
2,"HGB plain (no weighting, no cats)",41.20,60.50,1.75,91.95
3,"RandomForest (sqrt, distance-weighted, n=100)",43.53,62.79,1.97,94.23
4,"ExtraTrees (sqrt, distance-weighted, n=100)",48.98,71.29,2.17,99.61
5,"ExtraTrees (sqrt, n=100)",49.11,72.19,2.12,99.75


### Benchmark Takeaways

| Observation | Detail |
|---|---|
| **HGB family dominates** | All HGB variants beat all tree ensemble variants (ET, RF) on clean MAE |
| **Distance weighting helps** | HGB plain -> HGB-weighted: **\$41.20 -> \$39.96** (-\$1.24). Penalising long-haul dollar variance during training reduces error |
| **Native categoricals help** | DollarHGB vs HGB-weighted: **\$39.28 vs \$39.96** (-\$0.68). `pickup_cat/delivery_cat/equipment_cat` via `from_dtype` outperform numeric encoding alone |
| **Duan smearing** | S=1.0001 -- near-zero correction here (well-calibrated residuals), but guards against log-transform bias with zero downside risk |
| **ExtraTrees (old winner) now 5th** | Was best on older feature set without distance weighting. With the new preprocessing + weighting, HGB is clearly superior |

**Conclusion:** `DollarHGBRegressor` is the production model (\$39.28 clean MAE, 1.69% MAPE, 3MB vs 800MB).

## 4. DollarHGBRegressor Details
`HistGradientBoostingRegressor(max_iter=300, leaves=63, lr=0.05, from_dtype)` ~3MB vs 800MB. Unseen cities/equipment -> NaN missing-bin + `lane_freq==0`.

In [5]:
model = T.make_model()
print(type(model).__name__, model.get_params())
import joblib
art = Path('../output/pipeline.joblib')
print('artifact exists:', art.is_file(), ('%.2f MB' % (art.stat().st_size/1e6)) if art.is_file() else '')

DollarHGBRegressor {'max_iter': 300, 'max_leaf_nodes': 63, 'learning_rate': 0.05, 'min_samples_leaf': 20, 'l2_regularization': 1.0, 'categorical_features': 'from_dtype', 'early_stopping': False, 'random_state': 42}
artifact exists: True 3.12 MB


## 5. Dollar Calibration -- `sample_weight=distance`
Unweighted log-MAE over-weights short hauls; business pays dollars (`error x distance`). Weighting by distance penalises long-haul dollar variance during HGB training.

In [6]:
p_w = T.fit_predict(trf, vaf)  # production path: weighted + smeared (see modeling.py)
from train import _flag_params, _apply_flag
c, md, mad_v = _flag_params(trf)
m = _apply_flag(vaf, c, md, mad_v)
print('Fold2 weighted+smeared:', T.score(vaf.posted_rate, p_w, m))

Fold2 weighted+smeared: {'clean_mae': 39.28, 'clean_rmse': 57.93, 'clean_mape_pct': 1.69, 'all_row_mae': 90.08}


## 6. Duan Smearing -- `S=mean(exp(resid))`
`E[rate]=distance*exp(mu)*E[exp(resid)]`. Naive `exp(mu)*distance` under-predicts; we return `mu+log(S)`
so downstream `exp*distance` is unbiased. Residuals clipped `[0.2,5.0]`, `S` clamped `[0.9,1.5]`.

In [7]:
prep2 = T.fit_preprocessors(trf)
atr = T.transform(trf, prep2)
mask = T._train_mask(trf)
cl = atr[~mask]
y_cl = np.log(cl.posted_rate / cl.distance)
w_cl = cl.distance.to_numpy()
hgb_fit = DollarHGBRegressor().fit(cl[MODEL_FEATURES], y_cl, sample_weight=w_cl)
resid = y_cl.to_numpy() - (hgb_fit.predict(cl[MODEL_FEATURES]) - hgb_fit.log_smearing_)
S = float(np.clip(np.exp(resid), .2, 5.0).mean())
print(f'fitted S={hgb_fit.smearing_:.4f} (recomputed {S:.4f}); logS={hgb_fit.log_smearing_:.5f}')
fig, ax = plt.subplots(figsize=(9,4))
ax.hist(resid, bins=120)
ax.set_title(f'Log-RPM residuals (Fold2 train, S={hgb_fit.smearing_:.4f})')
ax.set_xlabel('resid')
fig.savefig(RESULTS/'residuals.png', dpi=150, bbox_inches='tight')
plt.show()
plt.close(fig)

fitted S=1.0001 (recomputed 1.0001); logS=0.00010


## 7. Forward-chaining CV + Ablations (all 4 folds via `T.evaluate`)

In [8]:
metrics = T.evaluate(train_raw)
flat = {k: v for k, v in metrics.items() if k != 'ablation_market_index'}
display = pd.DataFrame({k: {m: v for m, v in r.items() if m != 'held_out_cities'} for k, r in flat.items()}).T
print(display.to_string())

                               clean_mae  clean_rmse  clean_mape_pct  all_row_mae
Fold 1 (Jul-Aug)                   43.09       62.55            1.84        95.23
Fold 2 (Aug-Sep)                   39.28       57.93            1.69        90.08
Fold 3 (Sep-Oct)                   66.05       90.25            2.84       121.81
Fold 4 (Oct only)                  66.71       89.71            2.93       126.86
Unseen-city holdout (Sep-Oct)      84.25      106.88            3.51       145.36


## 8. Stratified Slices (same framework as `score.slice_report`)
Equipment, distance deciles, seen vs unseen cities/lanes, holiday Dec 24-26/31.

In [9]:
pred_path = Path('../output/validation_predictions.csv')
if pred_path.is_file():
    preds = pd.read_csv(pred_path)
    print(f'Loaded scored validation predictions ({len(preds)} rows)')
    rep = slice_report(validation, preds, None, train)
    print(rep)
else:
    print('no scored predictions yet -- run T.main()')

Loaded scored validation predictions (12000 rows)


## 9. Feature Importances & Market Index Ablation Plots

In [10]:
from sklearn.inspection import permutation_importance
sample = cl.sample(min(4000, len(cl)), random_state=42)
r = permutation_importance(
    hgb_fit, sample[MODEL_FEATURES], np.log(sample.posted_rate/sample.distance),
    n_repeats=3, random_state=42, n_jobs=1)
imp = pd.Series(r.importances_mean, index=MODEL_FEATURES).sort_values(ascending=False).head(15)
fig, ax = plt.subplots(figsize=(9,5))
imp.sort_values().plot.barh(ax=ax)
ax.set_title('Permutation importance (Dollar-HGB, Fold2 train sample)')
fig.savefig(RESULTS/'feature_importances.png', dpi=150, bbox_inches='tight')
plt.show()
plt.close(fig)

ab = metrics['ablation_market_index']
rows_ab = []
for variant in ['with_market_index', 'without_market_index', 'denoised_daily_median']:
    for fold, v in ab[variant].items():
        rows_ab.append({'variant': variant, 'fold': fold, 'clean_mae': v['clean_mae']})
comp = pd.DataFrame(rows_ab).pivot(index='fold', columns='variant', values='clean_mae')
fig, ax = plt.subplots(figsize=(10,4))
comp.plot.bar(ax=ax)
ax.set_title('Market index ablation: clean MAE by fold (DollarHGB)')
ax.set_ylabel('$')
fig.savefig(RESULTS/'stage2_comparison.png', dpi=150, bbox_inches='tight')
plt.show()
plt.close(fig)
print('wrote feature_importances.png and stage2_comparison.png')

wrote feature_importances.png and stage2_comparison.png


## 10. Final Fit + Deliverables
`T.main()` refits on all clean training data, writes `output/validation_predictions.csv`,
`output/december_predictions.csv`, `metrics.json`, and 3MB `pipeline.joblib`.

In [11]:
T.main()

Corrupt labels flagged: 677 (1.41%)
Fold 1 (Jul-Aug) {'clean_mae': 43.09, 'clean_rmse': 62.55, 'clean_mape_pct': 1.84, 'all_row_mae': 95.23}
Fold 2 (Aug-Sep) {'clean_mae': 39.28, 'clean_rmse': 57.93, 'clean_mape_pct': 1.69, 'all_row_mae': 90.08}
Fold 3 (Sep-Oct) {'clean_mae': 66.05, 'clean_rmse': 90.25, 'clean_mape_pct': 2.84, 'all_row_mae': 121.81}
Fold 4 (Oct only) {'clean_mae': 66.71, 'clean_rmse': 89.71, 'clean_mape_pct': 2.93, 'all_row_mae': 126.86}
Unseen-city holdout (Sep-Oct) {'clean_mae': 84.25, 'clean_rmse': 106.88, 'clean_mape_pct': 3.51, 'all_row_mae': 145.36}
Validation mean $2046.93; December mean $796.61
Smearing S=1.0001; global_fallback=0.0000; unseen_lane=0.122
Wrote validation_predictions.csv, output/validation_predictions.csv, output/december_predictions.csv, metrics.json, pipeline.joblib
